firmware/thruna/DCM/Src/App/App_Regen.c

```diff
@@ -0,0 +1,160 @@
+#include "App_Regen.h"
+#include <stdlib.h>
+#include "math.h"
+
+/**
+ * Check if left or right wheel is greater than 5.0km/hr
+ * @param RegenBraking struct to populate data
+ * @return true if wheel speed meets this condition, false
+ * otherwise
+ */
+static bool wheel_speed_in_range(RegenBraking *regenAttr);
+
+/**
+ * Check if battery cells are less than 4.0V
+ * @param RegenBraking struct to populate data
+ * @return true battery cells meet this condition,
+ * false otherwise
+ */
+static bool power_limit_check(RegenBraking *regenAttr);
+
+/**
+ * Algorithm to send negative torque request dependent
+ * on accelerator pedal percentage in range [-50,50]
+ * and to do active differential or not
+ * @param inputs are inputs for active differential
+ * @param regenAttr struct provides for torque request
+ */
+static void compute_regen_torque_request(ActiveDifferential_Inputs *inputs, RegenBraking *regenAttr);
+
+static RegenBraking              regenAttributes = { .enable_active_differential = true };
+static ActiveDifferential_Inputs activeDifferentialInputs;
+
+const float max_regen_nm              = -50.0f;
+const float wheel_speed_threshold_kph = 5.0f;
+const float max_pedal_position        = -50.0f;
+const float max_torque_delta          = 10.0f;
+const float min_scaling_speed_kph     = 35.0f;
+
+void App_Run_Regen(bool *regen, float *prev_torque_request, float accelerator_pedal_percentage)
+{
+    activeDifferentialInputs.accelerator_pedal_percentage = accelerator_pedal_percentage;
+
+    if (accelerator_pedal_percentage >= 0)
+    {
+        return;
+    }
+    else if (App_Regen_Safety(&regenAttributes))
+    {
+        regenAttributes.prev_torque_request_Nm = *prev_torque_request;
+        activeDifferentialInputs.steering_angle_deg =
+            App_CanRx_FSM_SteeringAngle_Get() * APPROX_STEERING_TO_WHEEL_ANGLE;
+
+        compute_regen_torque_request(&activeDifferentialInputs, &regenAttributes);
+        *regen = true;
+    }
+    else
+    {
+        regenAttributes.left_inverter_torque_Nm  = 0.0;
+        regenAttributes.right_inverter_torque_Nm = 0.0;
+        regenAttributes.prev_torque_request_Nm   = 0.0;
+
+        App_CanTx_DCM_Warning_RegenNotAvailable_Set(true);
+    }
+
+    *prev_torque_request = regenAttributes.prev_torque_request_Nm;
+
+    App_Regen_Activate(regenAttributes.left_inverter_torque_Nm, regenAttributes.right_inverter_torque_Nm);
+}
+
+bool App_Regen_Safety(RegenBraking *regenAttr)
+{
+    bool battery_temp_in_range = App_CanRx_BMS_MaxCellTemperature_Get() < 45;
+    return battery_temp_in_range && wheel_speed_in_range(regenAttr) && power_limit_check(regenAttr);
+}
+
+void App_Regen_Activate(float left, float right)
```

Comment:
kind of a misleading name for what it is doing

Reply:
i'll change it